### PR TITLE
Add Clang support to Make build system

### PR DESCRIPTION
**Pull request by luismarques**
The first commit adds Clang support, by detecting the correct target prefix.

The other two commits address the large number of warning messages that were being generated when compiling with Clang. The first of them deals with the use of options unsupported by Clang, the other with the use of linker options during non-linking build steps.

### CODE REVIEW

File: Configuration.mk [hunk 1]
```diff
@@ -130,6 +134,10 @@ endif
 TOOLCHAIN_rv32imac := $(TOOLCHAIN_rv32i)
 TOOLCHAIN_rv32imc := $(TOOLCHAIN_rv32i)
 
+ifeq ($(findstring -gcc,$(CC)),-gcc)
+    CC_IS_GCC := yes
+endif
```
[the remark this answers is not in the document]
Comment (bradjc) [reply Configuration.mk:139]: If we are going to add support for clang and gcc, and document it, let's just get the interface we want in this PR.

File: AppMakefile.mk [hunk 2]
```diff
@@ -141,7 +141,7 @@ $$(BUILDDIR)/$(1)/$(2).custom.ld: $$(LAYOUT) | $$(BUILDDIR)/$(1)
 # Collect all desired built output.
 $$(BUILDDIR)/$(1)/$(2).elf: $$(OBJS_$(1)) $$(LIBS_$(1)) $$(LEGACY_LIBS_$(1)) $$(BUILDDIR)/$(1)/$(2).custom.ld | $$(BUILDDIR)/$(1)
 	$$(TRACE_LD)
-	$$(Q)$$(TOOLCHAIN_$(1))$$(CC) $$(CFLAGS) $$(CPPFLAGS) $$(CPPFLAGS_$(1))\
+	$$(Q)$$(TOOLCHAIN_$(1))$$(CC) $$(CFLAGS) $$(CPPFLAGS) $$(CPPFLAGS_$(1)) $$(WLFLAGS) $$(WLFLAGS_$(1))\
```
[the remark this answers is not in the document]
Comment (ppannuto) [reply AppMakefile.mk:149]: Yeah, sorry, I wasn't clear there -- I meant to add something that would visible when the runtime warning printed. I just pushed a commit.

File: Configuration.mk (hunk 1 above)
[the remark this answers is not in the document]
Comment (ppannuto) [reply Configuration.mk:139]: +1 to this

**Review by ppannuto — COMMENTED** [review]
Sorry for the longer series of changes here; we try pretty hard to make tooling ergonomic for new and inexperienced users, so we're a bit pickier here.

I think if we update to fix our (non-standard, my fault :( ) `-clang` vs `clang`, we should be good to go here.

**Comment by luismarques** [comment]
> I think if we update to fix our (non-standard, my fault :( ) `-clang` vs `clang`, we should be good to go here.

Do you want me to submit a patch for that? That feels like it should be a separate PR, and this one should be rebased on that, but maybe doing that splitting is just busywork.

@bradjc had also commented about updating the README to note about Clang support and how to get the toolchain. Should I leave that to the regular contributors? I have no idea what you want to propose regarding how to get a RISC-V-compatible Clang/LLVM-based toolchain.

**Comment by ppannuto** [comment]
Can leave the `-` fix to a later PR, please just open an issue so we don't lose track of the desired change.

Re README: I would imagine something very similar to step 4 from https://github.com/tock/libtock-c#prerequisites -- noting it's optional, but here's the commands you'd run on the major platforms

**Comment by bradjc** [comment]
> Can leave the `-` fix to a later PR, please just open an issue so we don't lose track of the desired change.

I disagree. Unless there is something particularly difficult about adding it let's make the change now rather than add more work to do later. 

>  updating the README to note about Clang support and how to get the toolchain.

At the very least we need to update the readme to document how to build using clang. As for getting the toolchain, you should at least document how you got yours.

**Review by bradjc — CHANGES_REQUESTED** [review]
After more thinking about this, we specifically use GCC because it supports the variant of PIC we need for userspace apps on cortex-m apps. I assume clang still does not have the needed features for our PIC use case?

If that is the case, then using clang is going to break all cortex-m apps, and it's not clear why to me our makefile would make it easy to support a compiler that breaks apps. Is the intent to only use this for RISC-V (where we don't have the correct support on GCC either)? If so, that is more reasonable, but we need to somehow feature gate this to only work for RISC-V builds. We want to avoid allowing unsuspecting users to build broken apps with our officially support make scripts.

**Review by bradjc — APPROVED** [review]
Ok! I pushed two commits which I think take care of all of my raised issues.

Summary:

- Added support for `make CLANG=1 RISCV=1`. Since clang is not supported on cortex-m targets, `make CLANG=1` alone doesn't do anything.
- Added new `CPPFLAGS_gcc`/`CFLAGS_gcc` variable, and all of the supporting variables to make that work on a per-architecture-family basis.
- Removed `$(CC)` entirely as there is no assumed global compiler.
- Converted some if statements to our convention of grouping variables.

I also looked at the output of `make V=1` to ensure that the gcc only CPP and C flags were getting passed correctly.

**Comment by bradjc** [comment]
I don't understand why the github actions didn't run.

**Comment by hudson-ayers** [comment]
In libtock-c Github Actions is configured to only run on the merge commit of the PR and master, so it can't run while there is a merge conflict.

**Review by alistair23 — APPROVED** [review]

**Comment by luismarques** [comment]
> After more thinking about this, we specifically use GCC because it supports the variant of PIC we need for userspace apps on cortex-m apps. I assume clang still does not have the needed features for our PIC use case?
> 
> If that is the case, then using clang is going to break all cortex-m apps, and it's not clear why to me our makefile would make it easy to support a compiler that breaks apps. Is the intent to only use this for RISC-V (where we don't have the correct support on GCC either)? If so, that is more reasonable, but we need to somehow feature gate this to only work for RISC-V builds. We want to avoid allowing unsuspecting users to build broken apps with our officially support make scripts.

I see that you have already addressed this with additional commits, but in any case here's the answer to your questions. I suppose that people could who want to use Clang to target Tock Arm could still use it today (after adjusting build settings) if they are OK with getting non-relocatable apps, so in principle that could be supported. Personally, I don't care about that use case and I understand if you also don't want to support it at the moment. For getting relocatability with Clang, I have produced a [proof-of-concept demo](github.com/lowRISC/epic-tock-c-example) of relocatable Tock apps, using an embedded PIC (ePIC) spec that we (lowRISC) are working on. That spec/implementation is for RISC-V but it could probably be straightforwardly adapted and ported for Arm, so if our proposal is well received by the LLVM community maybe that will create further motivation for bringing similar support for the Clang/LLVM Arm targets.

> At the very least we need to update the readme to document how to build using clang. As for getting the toolchain, you should at least document how you got yours.

I work on LLVM, so my workflows for producing toolchains are a bit developer-centric and probably not what you would recommend for most users. I just googled for Clang RISC-V toolchains and I saw that there are a few offerings but I don't have experience with those, so I don't have any particular recommendation.

**Comment by luismarques** [comment]
I checked the updated commit stack and it LGTM. Thanks for the additional contributions.

**Comment by bradjc** [comment]
@ppannuto

File: Configuration.mk [hunk 3]
```diff
@@ -147,13 +172,31 @@ override CPPFLAGS += \
...
+ifeq ($(findstring -gcc,$(CC_cortex-m)),-gcc)
+  override CPPFLAGS_toolchain_cortex-m += $(CPPFLAGS_gcc)
+  override CFLAGS_toolchain_cortex-m += $(CFLAGS_gcc)
+endif
+
+ifeq ($(findstring -gcc,$(CC_rv32)),-gcc)
+  override CPPFLAGS_toolchain_rv32 += $(CPPFLAGS_gcc)
+  override CFLAGS_toolchain_rv32 += $(CFLAGS_gcc)
+endif
+
```
Comment (ppannuto) [inline Configuration.mk:199]:
```suggestion

# note: There are no non-gcc, clang-specific flags currently in use, so there is no
# equivalent CPPFLAGS_clang currently. If there are clang-only flags in the future,
# one can/should be added.

```

**Review by ppannuto — APPROVED** [review]
Pushed two small changes, a comment and a version print that would hopefully help with stuff like #262 in the future. LGTM

**Comment by bradjc** [comment]
Merging this now as it is at the very least a big step in the right direction. If there are lingering concerns we can fix those in follow up PRs.

bors r+

**Pull request merged**